rs/nervous_system/histogram/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//! An "event" has some real number associated with it. For example, a request
+//! takes 123 ms to process.
+//!
+//! The Histogram struct in this crate does two things:
+//!
+//! 1. Groups events based on
+//!     a) the real number associated with the event
+//!     b) binning scheme
+//! 2. Counts how many events there are in each group.
+//!
+//! This is designed to plug into ic_metrics_encoder::MetricsEncoder::histogram_vec.
+//!
+//! A limitation of this is that instead of being able to use f64 for the real
+//! value associated with an event, you must use i64. You can usually work with
+//! this by picking a sufficiently small unit.
+
+use ic_metrics_encoder::LabeledHistogramBuilder;
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use std::collections::BTreeMap;
+
+#[cfg(test)]
+mod tests;
+
+lazy_static! {
+    /// As the name suggests, this is useful when constructing a Histogram.
+    ///
+    /// The value of this is vec![1,2,3, ... 10, 20, 30, ... 100, 125, 150, 175,
+    /// 200, 225, 250, 275, ...] and goes as high as i64 allows.
+    pub static ref STANDARD_POSITIVE_BIN_INCLUSIVE_UPPER_BOUNDS: Vec<i64> = {
```

Comment:
I guess my main point was to make the steps less arbitrary, e.g., the distance between 1, 2, 5, 10, 20, 50.., is always 2X or 2.5X times what we had in the previous step. 

Maybe other use cases could just use a linear scale of buckets? What's the highest number of buckets that you think is still reasonable to have?

Reply:
Not sure why what I have seems arbitrary. The main thing I was doing was taking steps that powers of ten. Admittedly, I made it 4x as fine grained at larger magnitudes, but the idea is the same.

Reply:
> 4x as fine grained at larger magnitudes

That's exactly why it seems arbitrary to me. 

Without knowing the nature of the thing being measured, it's unclear why some ranges are given preference over others (in terms of resolution). Maybe if you called these buckets in a more specific way, it'd be less controversial ("metric_x_buckets" or something like that), as then one could argue that these exact buckets are a good choice for metric X.

But for a general set of buckets, I'd just use either a linear scale or a logarithmic scale. 

That being said, I like that the core part of the PR does not prescribe which buckets to use, so this is a minor issue.

Reply:
IMHO, this is WAY TOO MANY buckets. Note that given Prometheus' cumulative histograms, this will result in hundreds of time series per histogram (one per bucket, whether or not there were any observations in that bucket). So e.g. even if all the values you observed were either `0` or `1`, you would still get 300+ time series.

Which brings up another relevant aspect: the observed values of virtually any metric I've seen tend to be clustered together (whether they're in the low integers or in the hundreds of GB), so if one is willing to put in a bit of effort, it's generally quite easy to come up with buckets that cover a quite limited range without any (or any meaningful) loss of precision.

Finally, giving everyone a choice between "limited range at limited resolution at the cost of extra brainpower and code, at the risk of having a finger pointed at me if I somehow miss out on an edge case" vs "any possible value at minute resolution for free" is predictably going to nudge people to just use this simply because it's there.

The reason why a histogram of this magnitude is a problem even in the context of single canisters (as opposed to hundreds and thousands of replicas) is that (for someone not really aware of what `STANDARD_POSITIVE_BINS` is or how it fits together with metrics and labels and whatnot, it's easy to think "well, I'm just going to add yet another dimension to my `HistogramVec`, since it only has 7 so far" (I've actually seen histograms with about that number of dimensions; and people wondering why their dashboard, that used to be quite snappy would no longer load). And if you already have 300+ buckets per label combination, any new dimension can potentially mean tens to hundreds of thousands of time series. For a metric that maybe no one really cares enough to add to a dashboard to begin with.

The common approach in the replica is to use `binary_buckets()` (approximate powers of 2, with 2.5 thrown in to make them nice, round numbers) plus specific values inserted here and there because they're meaningful (e.g. if 7B instructions is some limit used by Execution, then it makes sense to have a 7B bucket in-between 5B and 10B; or if values in the billions tend to be problematic, then insert 3B, 4B, 6B, etc. buckets in addition to the existing powers of 2; but likely no one cares about 3 instructions; or 3.25^15 instructions).

Reply:
Thanks. I think the tl;dr is that this is too many. I will get rid of this.